mllib/src/main/scala/org/apache/spark/mllib/stat/test/KSTest.scala

```diff
@@ -0,0 +1,126 @@
+
+import org.apache.spark.{SparkException, Logging}
+import org.apache.spark.rdd.RDD
+
+
+/**
+ * Conduct the two-sided Kolmogorov Smirnov test for data sampled from a
+ * continuous distribution. By comparing the largest difference between the empirical cumulative
+ * distribution of the sample data and the theoretical distribution we can provide a test for the
+ * the null hypothesis that the sample data comes from that theoretical distribution.
+ * For more information on KS Test: https://en.wikipedia.org/wiki/Kolmogorov%E2%80%93Smirnov_test
+ */
+  private[stat] object KSTest {
+
+  // Null hypothesis for the type of KS test to be included in the result.
+  object NullHypothesis extends Enumeration {
+    type NullHypothesis = Value
+    val oneSampleTwoSided = Value("Sample follows theoretical distribution.")
+  }
+
+  /**
+   * Calculate empirical cumulative distribution values needed for KS statistic
+   * @param dat `RDD[Double]` on which to calculate empirical cumulative distribution values
+   * @return and RDD of (Double, Double, Double), where the first element in each tuple is the
+   *         value, the second element is the ECDFV - 1 /n, and the third element is the ECDFV,
+   *         where ECDF stands for empirical cumulative distribution function value
+   *
+   */
+  def empirical(dat: RDD[Double]): RDD[(Double, Double, Double)] = {
+    val n = dat.count().toDouble
+    dat.sortBy(x => x).zipWithIndex().map { case (v, i) => (v, i / n, (i + 1) / n) }
+  }
+
+  /**
+   * Runs a KS test for 1 set of sample data, comparing it to a theoretical distribution
+   * @param dat `RDD[Double]` to evaluate
+   * @param cdf `Double => Double` function to calculate the theoretical CDF
+   * @return a KSTestResult summarizing the test results (pval, statistic, and null hypothesis)
+   */
+  def testOneSample(dat: RDD[Double], cdf: Double => Double): KSTestResult = {
+    val empiriRDD = empirical(dat) // empirical distribution
+    val distances = empiriRDD.map {
+        case (v, dl, dp) =>
+          val cdfVal = cdf(v)
+          Math.max(cdfVal - dl, dp - cdfVal)
+      }
+    val ksStat = distances.max()
+    evalOneSampleP(ksStat, distances.count())
+  }
+
+  /**
+   * Runs a KS test for 1 set of sample data, comparing it to a theoretical distribution. Optimized
+   * such that each partition runs a separate mapping operation. This can help in cases where the
+   * CDF calculation involves creating an object. By using this implementation we can make sure
+   * only 1 object is created per partition, versus 1 per observation.
+   * @param dat `RDD[Double]` to evaluate
+   * @param distCalc a function to calculate the distance between the empirical values and the
+   *                 theoretical value
+   * @return a KSTestResult summarizing the test results (pval, statistic, and null hypothesis)
+   */
+  def testOneSampleOpt(dat: RDD[Double],
+                       distCalc: Iterator[(Double, Double, Double)] => Iterator[Double])
```

Comment:
`distCalc` should be indented four spaces past `def` and `: KSTestResult` should be indented two spaces past `def`